AppCheck + AppAttest getAppCheckToken failing for 10% of users in production

### Description

We rolled out Firebase AppCheck to our iOS users last month, however there are a significant minority of users in production who are getting issues from DeviceCheck:

```
Failed to create a block of data that demonstrates the legitimacy of the app instance (`generateAssertion:clientDataHash:completionHandler:`); keyId.length = 44, clientDataHash.length = 32; DCErrorInvalidInput - An error code that indicates when your app provides data that isn’t formatted correctly.
--

(Error Domain=com.apple.devicecheck.error Code=2 "(null)")
```

Our AppCheck implementation is as defined in the docs:

```
func getAppCheckToken() async throws -> String {
        try await AppCheck.appCheck().token(forcingRefresh: false).token    
}
```

There was an issue last year, in May 2023, somebody with the exact same problem, getting frequent device check errors (`DCErrorInvalidInput`), as well as some less frequent errors  (`DCErrorInvalidKey`).

https://github.com/firebase/firebase-ios-sdk/issues/11264

In November 2023, a change was rolled out in 10.18.0 which claimed to fix the `DCErrorInvalidKey` issue, however it didn't mention the far more common `DCErrorInvalidInput` issue. 

https://github.com/firebase/firebase-ios-sdk/pull/11986 


### Reproducing the issue

Unsure specifically what's causing the issue but rolling out AppCheck into production caught a large minority of users who have this issue.

### Firebase SDK Version

10.23.0

### Xcode Version

15.2

### Installation Method

Swift Package Manager

### Firebase Product(s)

App Check

### Targeted Platforms

iOS

### Relevant Log Output

```shell
nserror-domain:	
com.firebase.appCheck

NSLocalizedFailureReason:	
Failed to create a block of data that demonstrates the legitimacy of the app instance (`generateAssertion:clientDataHash:completionHandler:`); keyId.length = 44, clientDataHash.length = 32; DCErrorInvalidInput - An error code that indicates when your app provides data that isn’t formatted correctly.

NSUnderlyingError:	
Error Domain=com.apple.devicecheck.error Code=2 "(null)"
```


### If using Swift Package Manager, the project's Package.resolved

_No response_

### If using CocoaPods, the project's Podfile.lock

_No response_

@andrewheard we are using 10.22 the following scenario is reproducible for us every time.

User is logged in and is issued AppCheck and Firebase token on iPhone X, user acquires a new phone and onboards with the same Apple ID to the new phone, electing to "restore from iCloud" to the new device

on opening the app for the first time, we get the following logs. All further app check tokens fails.
When deleting the app from the new iPhone and reinstalling, the issue dissappears

`2024-04-24 19:55:22.941 [App 58FF] [INFO] [AppLifecycleManager] sceneWillEnterForeground(_:) (AppLifecycleManager.swift,163): Lifecycle: sceneWillEnterForeground 1DEA954B-9469-47CD-A4D3-1290CAEFD5E0
2024-04-24 19:55:22.943 [App 58FF] [INFO] [AppLifecycleManager] sceneWillEnterForeground(_:) (AppLifecycleManager.swift,171): no current user while entering foreground
2024-04-24 19:55:22.956 [App 58FF] [INFO] [AppLifecycleManager] appWillEnterForeground() (AppLifecycleManager.swift,104): Lifecycle: appWillEnterForeground
2024-04-24 19:55:23.240 [App 58FF] [INFO] [AppLifecycleManager] sceneDidBecomeActive(_:) (AppLifecycleManager.swift,179): Lifecycle: sceneDidBecomeActive 1DEA954B-9469-47CD-A4D3-1290CAEFD5E0
2024-04-24 19:55:23.241 [App 58FF] [INFO] [ApplicationBadgeUpdater] sceneDidBecomeActive() (ApplicationBadgeUpdater.swift,38): Zeroing application badge - sceneDidBecomeActive
2024-04-24 19:55:23.241 [App 58FF] [INFO] [ApplicationBadgeUpdater] updateApplicationBadge(to:) (ApplicationBadgeUpdater.swift,30): Setting application badge count to 0
2024-04-24 19:55:23.244 [App 58FF] [INFO] [AppLifecycleManager] appDidBecomeActive() (AppLifecycleManager.swift,108): Lifecycle: appDidBecomeActive
2024-04-24 19:55:24.436 [App 58FF] [ERROR] [AuthUtils] getAppCheckToken(completion:) (AuthUtils.swift,23): Failed to get AppCheck token with error: Optional(Error Domain=com.firebase.appCheck Code=0 "Failed to attest the validity of the generated cryptographic key (`attestKey:clientDataHash:completionHandler:`); keyId.length = 44, clientDataHash.length = 32; DCErrorInvalidInput - An error code that indicates when your app provides data that isn’t formatted correctly." UserInfo={NSLocalizedFailureReason=Failed to attest the validity of the generated cryptographic key (`attestKey:clientDataHash:completionHandler:`); keyId.length = 44, clientDataHash.length = 32; DCErrorInvalidInput - An error code that indicates when your app provides data that isn’t formatted correctly., NSUnderlyingError=0x303f889c0 {Error Domain=com.apple.devicecheck.error Code=2 "(null)"}})`

--- a/.github/workflows/check.yml
+++ b/.github/workflows/check.yml
@@ -17,10 +17,6 @@ jobs:
     steps:
     - uses: actions/checkout@v3
 
-    - uses: actions/setup-python@v3
-      with:
-        python-version: 3.6
-
     - name: Cache Mint packages
       uses: actions/cache@v3
       with:

--- a/AppCheckCore/Sources/AppAttestProvider/GACAppAttestProvider.m
+++ b/AppCheckCore/Sources/AppAttestProvider/GACAppAttestProvider.m
@@ -363,15 +363,19 @@ NS_ASSUME_NONNULL_BEGIN
               })
       .recoverOn(self.queue,
                  ^id(NSError *error) {
-                   // If Apple rejected the key (DCErrorInvalidKey) then reset the attestation and
+                   // If Apple rejected the key (DCErrorInvalidKey or
+                   // DCErrorInvalidInput) then reset the attestation and
                    // throw a specific error to signal retry (GACAppAttestRejectionError).
                    NSError *underlyingError = error.userInfo[NSUnderlyingErrorKey];
                    if (underlyingError && [underlyingError.domain isEqualToString:DCErrorDomain] &&
-                       underlyingError.code == DCErrorInvalidKey) {
-                     GACAppCheckLog(
-                         GACLoggerAppCheckMessageCodeAttestationRejected, GACAppCheckLogLevelDebug,
-                         @"App Attest invalid key; the existing attestation will be reset.");
-
+                       (underlyingError.code == DCErrorInvalidKey ||
+                        underlyingError.code == DCErrorInvalidInput)) {
+                     NSString *logMessage = [NSString
+                         stringWithFormat:@"App Attest invalid key/input; the existing attestation "
+                                          @"will be reset. DC Error Code: %@.",
+                                          @(underlyingError.code)];
+                     GACAppCheckLog(GACLoggerAppCheckMessageCodeAttestationRejected,
+                                    GACAppCheckLogLevelDebug, logMessage);
                      // Reset the attestation.
                      return [self resetAttestation].thenOn(self.queue, ^NSError *(id result) {
                        // Throw the rejection error.

--- a/AppCheckCore/Tests/Unit/AppAttestProvider/GACAppAttestProviderTests.m
+++ b/AppCheckCore/Tests/Unit/AppAttestProvider/GACAppAttestProviderTests.m
@@ -628,73 +628,16 @@ GAC_APP_ATTEST_PROVIDER_AVAILABILITY
 }
 
 - (void)testGetToken_WhenExistingKeyIsRejectedByApple_ThenAttestationIsResetAndRetriedOnce_Success {
-  // 1. Expect GACAppAttestService.isSupported.
-  [OCMExpect([self.mockAppAttestService isSupported]) andReturnValue:@(YES)];
-
-  // 2. Expect storage getAppAttestKeyID.
-  NSString *existingKeyID = @"existingKeyID";
-  OCMExpect([self.mockStorage getAppAttestKeyID])
-      .andReturn([FBLPromise resolvedWith:existingKeyID]);
-
-  // 3. Expect a stored artifact to be requested.
-  __auto_type rejectedPromise = [self rejectedPromiseWithError:[NSError errorWithDomain:self.name
-                                                                                   code:NSNotFound
-                                                                               userInfo:nil]];
-  OCMExpect([self.mockArtifactStorage getArtifactForKey:existingKeyID]).andReturn(rejectedPromise);
-
-  // 4. Expect random challenge to be requested.
-  OCMExpect([self.mockAPIService getRandomChallenge])
-      .andReturn([FBLPromise resolvedWith:self.randomChallenge]);
-
-  // 5. Expect the key to be attested with the challenge.
-  NSError *attestationError = [NSError errorWithDomain:DCErrorDomain
-                                                  code:DCErrorInvalidKey
-                                              userInfo:nil];
-  id attestCompletionArg = [OCMArg invokeBlockWithArgs:[NSNull null], attestationError, nil];
-  OCMExpect([self.mockAppAttestService attestKey:existingKeyID
-                                  clientDataHash:self.randomChallengeHash
-                               completionHandler:attestCompletionArg]);
-
-  // 6. Stored attestation to be reset.
-  [self expectAttestationReset];
-
-  // 7. Expect the App Attest key pair to be generated and attested.
-  NSString *newKeyID = @"newKeyID";
-  NSData *attestationData = [[NSUUID UUID].UUIDString dataUsingEncoding:NSUTF8StringEncoding];
-  [self expectAppAttestKeyGeneratedAndAttestedWithKeyID:newKeyID attestationData:attestationData];
-
-  // 8. Expect exchange request to be sent.
-  GACAppCheckToken *appCheckToken = [[GACAppCheckToken alloc] initWithToken:@"App Check Token"
-                                                             expirationDate:[NSDate date]];
-  NSData *artifactData = [@"attestation artifact" dataUsingEncoding:NSUTF8StringEncoding];
-  __auto_type attestKeyResponse =
-      [[GACAppAttestAttestationResponse alloc] initWithArtifact:artifactData token:appCheckToken];
-  OCMExpect([self.mockAPIService attestKeyWithAttestation:attestationData
-                                                    keyID:newKeyID
-                                                challenge:self.randomChallenge
-                                               limitedUse:NO])
-      .andReturn([FBLPromise resolvedWith:attestKeyResponse]);
-
-  // 9. Expect the artifact received from Firebase backend to be saved.
-  OCMExpect([self.mockArtifactStorage setArtifact:artifactData forKey:newKeyID])
-      .andReturn([FBLPromise resolvedWith:artifactData]);
-
-  // 10. Call get token.
-  XCTestExpectation *completionExpectation =
-      [self expectationWithDescription:@"completionExpectation"];
-  [self.provider
-      getTokenWithCompletion:^(GACAppCheckToken *_Nullable token, NSError *_Nullable error) {
-        [completionExpectation fulfill];
-
-        XCTAssertEqualObjects(token.token, appCheckToken.token);
-        XCTAssertEqualObjects(token.expirationDate, appCheckToken.expirationDate);
-        XCTAssertNil(error);
-      }];
-
-  [self waitForExpectations:@[ completionExpectation ] timeout:0.5 enforceOrder:YES];
-
-  // 11. Verify mocks.
-  [self verifyAllMocks];
+  NSError *invalidKeyError = [NSError errorWithDomain:DCErrorDomain
+                                                 code:DCErrorInvalidKey
+                                             userInfo:nil];
+  [self assertAttestationResetAndGetTokenRetryWhenExistingKeyIsRejectedWithAttestationError:
+            invalidKeyError];
+  NSError *invalidInputError = [NSError errorWithDomain:DCErrorDomain
+                                                   code:DCErrorInvalidInput
+                                               userInfo:nil];
+  [self assertAttestationResetAndGetTokenRetryWhenExistingKeyIsRejectedWithAttestationError:
+            invalidInputError];
 }
 
 #pragma mark - FAC token refresh (assertion)
@@ -1175,6 +1118,74 @@ GAC_APP_ATTEST_PROVIDER_AVAILABILITY
   // 9. Verify backoff result.
   XCTAssertEqualObjects(((GACAppCheckToken *)self.fakeBackoffWrapper.operationResult).token,
                         FACToken.token);
+}
+
+- (void)assertAttestationResetAndGetTokenRetryWhenExistingKeyIsRejectedWithAttestationError:
+    (NSError *)error {
+  // 1. Expect GACAppAttestService.isSupported.
+  [OCMExpect([self.mockAppAttestService isSupported]) andReturnValue:@(YES)];
+
+  // 2. Expect storage getAppAttestKeyID.
+  NSString *existingKeyID = @"existingKeyID";
+  OCMExpect([self.mockStorage getAppAttestKeyID])
+      .andReturn([FBLPromise resolvedWith:existingKeyID]);
+
+  // 3. Expect a stored artifact to be requested.
+  __auto_type rejectedPromise = [self rejectedPromiseWithError:[NSError errorWithDomain:self.name
+                                                                                   code:NSNotFound
+                                                                               userInfo:nil]];
+  OCMExpect([self.mockArtifactStorage getArtifactForKey:existingKeyID]).andReturn(rejectedPromise);
+
+  // 4. Expect random challenge to be requested.
+  OCMExpect([self.mockAPIService getRandomChallenge])
+      .andReturn([FBLPromise resolvedWith:self.randomChallenge]);
+
+  // 5. Expect the key to be attested with the challenge.
+  id attestCompletionArg = [OCMArg invokeBlockWithArgs:[NSNull null], error, nil];
+  OCMExpect([self.mockAppAttestService attestKey:existingKeyID
+                                  clientDataHash:self.randomChallengeHash
+                               completionHandler:attestCompletionArg]);
+
+  // 6. Stored attestation to be reset.
+  [self expectAttestationReset];
+
+  // 7. Expect the App Attest key pair to be generated and attested.
+  NSString *newKeyID = @"newKeyID";
+  NSData *attestationData = [[NSUUID UUID].UUIDString dataUsingEncoding:NSUTF8StringEncoding];
+  [self expectAppAttestKeyGeneratedAndAttestedWithKeyID:newKeyID attestationData:attestationData];
+
+  // 8. Expect exchange request to be sent.
+  GACAppCheckToken *appCheckToken = [[GACAppCheckToken alloc] initWithToken:@"App Check Token"
+                                                             expirationDate:[NSDate date]];
+  NSData *artifactData = [@"attestation artifact" dataUsingEncoding:NSUTF8StringEncoding];
+  __auto_type attestKeyResponse =
+      [[GACAppAttestAttestationResponse alloc] initWithArtifact:artifactData token:appCheckToken];
+  OCMExpect([self.mockAPIService attestKeyWithAttestation:attestationData
+                                                    keyID:newKeyID
+                                                challenge:self.randomChallenge
+                                               limitedUse:NO])
+      .andReturn([FBLPromise resolvedWith:attestKeyResponse]);
+
+  // 9. Expect the artifact received from Firebase backend to be saved.
+  OCMExpect([self.mockArtifactStorage setArtifact:artifactData forKey:newKeyID])
+      .andReturn([FBLPromise resolvedWith:artifactData]);
+
+  // 10. Call get token.
+  XCTestExpectation *completionExpectation =
+      [self expectationWithDescription:@"completionExpectation"];
+  [self.provider
+      getTokenWithCompletion:^(GACAppCheckToken *_Nullable token, NSError *_Nullable error) {
+        [completionExpectation fulfill];
+
+        XCTAssertEqualObjects(token.token, appCheckToken.token);
+        XCTAssertEqualObjects(token.expirationDate, appCheckToken.expirationDate);
+        XCTAssertNil(error);
+      }];
+
+  [self waitForExpectations:@[ completionExpectation ] timeout:0.5 enforceOrder:YES];
+
+  // 11. Verify mocks.
+  [self verifyAllMocks];
 }
 
 - (void)expectAppAttestAvailabilityToBeCheckedAndNotExistingStoredKeyRequested {

